documentation/smg/17_pixel_data.md

```diff
@@ -0,0 +1,427 @@
+With this interface, Horace algorithms could access pixel data on disk
+in a similar way to which they access pixels in memory.
+The object could wrap the raw pixel data and/or a file handle,
+and provide getters/setters for the data.
+
+The concept of "pages" of data would exist in the `PixelData` class.
+Instances of the class could load, and operate on, a "page" of data at a time.
+The class would have a `has_more()` method to indicate whether data subsequent
+to the currently loaded page exists on disk.
+This would allow looping over pixel data in a `while` loop:
+
+```matlab
+while pixel_data.has_more()
+    % operate on pixel page
+end
+```
+
+### Data API
+
+Pixel data, in memory, was stored as a raw Matlab array.
+Pixel data fields (e.g. signal, variance etc.) were accessed using indexing:
+
+```matlab
+>> pix_signal = sqw_obj.data.pix(8, :);
+>> pix_coordinates = sqw_obj.data.pix(1:4, :);
+```
+
+This meant that accessing data required previous knowledge of the data structure;
+to get the signal, one needed to know it is stored in column 8.
+It also meant that actions performed on pixel data relied on the array's structure.
+This would cause issues if it was ever advantageous for the structure of the
+pixel data to be changed.
+
+#### Data API - Solution
+
+It was proposed that the `PixelData` class would provide getters/setters for
+pixel data fields.
+This would not assume users of the class had knowledge of the underlying data
+structure.
+It would also mean that, if the pixel data structure were to change,
+other code using pixel data could remain the same.
+
+### Extendability
+
+There was also a requirement that the pixel data be extendable,
+i.e. users should be able to add extra named columns to the data.
+
+#### Extendability - Solution
+
+The proposed `PixelData` class would allow addition of extra data per pixel,
+in named elements.
+These data could be retrieved using a getter that takes a char/string argument.
+
+### Parallelism
+
+As pixel data is often large, it's advantageous to perform operations on the
+data using distributed memory parallelism.
+To aid in this, a simple way of "chunking" the data between workers was required.
+
+#### Parallelism - Solution
+
+The ability to move to specific pages of pixel data was proposed to solve this
+problem.
+Using the `move_to_page(page_num)` method allows a worker to position itself at
+a particular point in the pixel data.
+This reduces distributing pixels between workers to a function of:
+the number of workers, the number of pixel, the page size.
+Each worker can receive its start and end page index from the head node,
+and operate on those pixels.
+
+## The PixelData Class
+
+### Principles
+
+- It operates with "pages" of cached data.
+  - The page size is specified on construction (in bytes).
+  - Only a page size worth of pixels is loaded at any one time.
+  - If all pixels fit in one page, all pixels are loaded.
+
+- Temporary files hold changes to pixel data.
```

Comment:
It would be helpful to include information on how to save an SQW file to merge temp files back and have an updated file on disk.
Does that work if you save back over the original file or does it require a new file to be created?